include/cetl/span.h

```diff
@@ -1 +1,772 @@
+    /// @see std::span::first
+    template <size_type Count>
+    constexpr span<element_type, Count> first() const
+    {
+        static_assert(Count <= Extent, "subviews beyond the size of the span's view are not allowed.");
+        return span<element_type, Count>(data_, Count);
+    }
+
+    /// Create a new span from the start of the current span for count elements.
+    /// @param count  The number of elements for the sub-span. Where `count` is greater than the
+    ///               current span's size the behavior is undefined.
+    /// @return A new span with a dynamic extent extent and a size of `count`.
+    /// @see std::span::first
+    constexpr span<element_type, cetl::dynamic_extent> first(size_type count) const
+    {
+        CETL_DEBUG_ASSERT(count <= Extent, "Dynamic subviews beyond the size of the span's view are undefined.");
+        return span<element_type, cetl::dynamic_extent>(data_, count);
+    }
+
+    /// Create a new span `Count` elements from the last item of the current span to its end.
+    /// @tparam Count   Number of elements for the sub-span.
+    /// @return A new span with the static extent `Count`.
+    /// @see std::span::last
+    template <size_type Count>
+    constexpr span<element_type, Count> last() const
+    {
+        static_assert(Count <= extent, "subviews beyond the size of the span's view are not allowed.");
+        return span<element_type, Count>{(data_ + extent) - Count, Count};
+    }
+
+    /// Create a new span `count` elements from the last item of the current span to its end.
+    /// @param count   Number of elements for the sub-span. The behavior of this method is undefined where
+    ///                `count` is greater than the current span's size.
+    /// @return A new span with a size of `count` and a dynamic extent.
+    /// @see std::span::last
+    constexpr span<element_type, cetl::dynamic_extent> last(size_type count) const
+    {
+        CETL_DEBUG_ASSERT(count <= extent, "Dynamic subviews beyond the size of the span's view are undefined.");
+        return span<element_type, cetl::dynamic_extent>((data_ + extent) - count, count);
+    }
+
+    /// Create a new span `Offset` elements from the start of the current span and for `Count` elements.
+    /// @tparam Offset  Number of elements from the start of the current span for the subspan.
+    /// @tparam Count   The number of elements from the Offset to include in the subspan. If this value is
+    ///                 `cetl::dynamic_extent` then the Count is `size() - Offset`.
+    /// @return A new span where the `Extent` is `Count` if this was not `cetl::dynamic_extent` otherwise the
+    ///         new span's `Extent` is this span's `Extent` minus `Offset`.
+    /// @see std::span::subspan()
+    template <size_type Offset, size_type Count = cetl::dynamic_extent>
+    constexpr auto subspan() const -> span<element_type, (Count != cetl::dynamic_extent) ? Count : Extent - Offset>
```

Comment:
Since this is suppose to be C++20 everywhere, should the -> notion be more common? I'm not sure I like it but consistency is nice.

Reply:
This is the only way I know how to implement this because the return type depends on the template parameters. Nowhere else, in this class, is this true.

Reply:
Why can't you move the return type declaration to the left? It is not dependent on the runtime arguments since there are none

Reply:
Huh. you're right. Somehow I convinced myself that this wouldn't work but...it totally does.